### PR TITLE
DEP: Deprecate private but non-underscored ndimage.<module> namespace

### PR DESCRIPTION
**Pull request by AnirudhDagar**
#### Reference issue
See gh-14360 for details. Follows gh-14419 pattern.

#### What does this implement/fix?
Adds deprecation warning to private but non-underscored ndimage modules.

Also, I've noticed that our documentation examples show the use of this private namespace for example in the `center_of_mass` method. See [`center_of_mass` docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.center_of_mass.html#scipy.ndimage.center_of_mass).

Therefore, I've replaced any such instance in documentation or internal calls of `ndimage.<module>.<some_func>` to `ndimage.<some_func>` avoiding deprecation warnings in our own code.

#### Additional information

**Note: This PR is planned to be merged after #14356**. It will need a very minor change (replace `_<module>` with `_multimethods`) after that merge, which I'll do by rebasing again. I wanted to get this up early, for feedback, if any. Hence raising this WIP PR before #14356 actually lands.

**Edit 1:**
Ps. Technically can go ahead with this before #14356 as well (then I can rebase that PR), but let's wait for some time anyways. :)

**Edit 2:**
Q. What about places like [this](https://github.com/scipy/scipy/blob/053f3f54130f8b17c09540352a784f4c54a919df/scipy/ndimage/_interpolation.py#L76) in the docs where we mention the private namespace? Should we also re-phrase these? Small things like these could actually contribute towards someone using a deprecated namespace.

cc @rgommers

### CODE REVIEW

**Comment by grlee77** [comment]
> Q. What about places like this in the docs where we mention the private namespace? Should we also re-phrase these? Small things like these could actually contribute towards someone using a deprecated namespace.

Yes, I would rephrase "All functions in `ndimage.interpolation`" to something like "All of the interpolation functions in `ndimage`"

**Comment by AnirudhDagar** [comment]
I've removed the WIP label here. I feel it is ready and we can merge this before #14356 (which still needs some fixes). I'll rebase the latter then.

File: scipy/ndimage/filters.py [hunk 1]
```diff
@@ -1,3 +1,8 @@
+# This file is not meant for public use and will be removed in the future
+# versions of SciPy. Use the `scipy.ndimage` namespace for importing the
```
Comment (grlee77) [inline scipy/ndimage/filters.py:2]: If we know what version we plan to remove this, I would mention the specific version here (and in the other files)

**Review by grlee77 — APPROVED** [review]
Not sure if it was already discussed in which version the removal would occur? If so I would add that as suggested in my comment here. If this is on an indefinite timescale then it looks good to me as is.

File: scipy/ndimage/filters.py (hunk 1 above)
Comment (grlee77) [reply scipy/ndimage/filters.py:2]: If we are not listing a specific version, I would just make a slight edit from "in the future versions" to "in a future version"

Comment (AnirudhDagar) [reply scipy/ndimage/filters.py:2]: @rgommers mentioned the expiry with SciPy 2.0 in the [message](https://mail.python.org/pipermail/scipy-dev/2021-July/024943.html) to the mailing list.

**Comment by rgommers** [comment]
+1 for saying removal in `2.0.0`, which is likely a very long time into the future.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @AnirudhDagar

**Comment by AnirudhDagar** [comment]
Ping @grlee77 @rgommers.
Should we merge this now? It's been open to comments, but I guess everyone's happy here. :)

File: scipy/ndimage/filters.py [hunk 2]
```diff
@@ -20,4 +25,8 @@ def __getattr__(name):
         raise AttributeError(
             f"scipy.ndimage.filters has no attribute {name}.")
```
Comment (peterbell10) [inline scipy/ndimage/filters.py:26]: You should put the deprecation message in the attribute error as well.

Comment (AnirudhDagar) [reply scipy/ndimage/filters.py:26]: Will that not be a little confusing for a function that actually doesn't exist at all?

If not, does this look ok to you?
```suggestion
            f"scipy.ndimage.filters has no attribute {name}. "
             "Please use the `scipy.ndimage` namespace, "
             "the `scipy.ndimage.filters` namespace is deprecated.")
```

Comment (peterbell10) [reply scipy/ndimage/filters.py:26]: >Will that not be a little confusing for a function that actually doesn't exist at all?

It's not the function that's deprecated, it's the module namespace. So, I don't think it's confusing at all.

I would be consistent with `signal.spline`'s message:
https://github.com/scipy/scipy/blob/e4b3e6eb372b8c1d875f2adf607630a31e2a609c/scipy/signal/spline.py#L19-L21

**Comment by AnirudhDagar** [comment]
@peterbell10 I've addressed your comment. The failing tests are probably unrelated.

**Review by peterbell10 — APPROVED** [review]

**Comment by rgommers** [comment]
Has had 3 approvals for a while and CI is all green now - in it goes. Thanks @AnirudhDagar! And thanks for the reviews @grlee77 and @peterbell10
